File: session06_GrowArray_LinkedList_Stack_Queue/GrowArray/BadGrowArray2018F.java
public class BadGrowArray2018F {
	private	int len;
	private int[] data;

	public BadGrowArray() {
		data = null; // new int[len]
		len = 0;
	}
	public void insertEnd(int v) {
		int[] old = data;
		data = new int[len+1];
		for (int i = 0; i < len; i++)
			data[i] = old[i];
		data[len] = v;
		len++;
	}
	
  public void insertStart(int v) {
		int[] old = data;
		data = new int[len+1];
		data[0] = v;
		for (int i = 0; i < len; i++)
			data[i+1] = old[i];
		len++;
	}

  public void insert(int i, int v) { //O(len)
		int[] old = data;
		data = new int[len+1];
		for (int j = 0; j < i; j++)
			data[j] = old[j];
		data[i] = v;
		for (int j = i; j < len; j++)
			data[j+1] = temp[j];
		len++;
	}

	
	// remove from end and return the value being removed
	public int removeEnd() {  //O(1)
		int temp = data[len-1];
		len--;
		return temp;
	}

	public int removeStart() { //O(len)
		int temp = data[0];
		--len;
		for (int i = 0; i < len; i++)
      data[i] = data[i+1];			
	}

  public int get(int pos)  {
		return data[pos];//O(1)
	}
	public int getSize() { //O(1)
		return len;
	}
	public String toString() {
		// StringBuilder b = new StringBuilder();
		// b.append( );
		// return b.toString();
		String s = "";
		for (int i = 0; i < length; i++)
			s += data[i] + ' ';
		return s;
	}
	
	public static void main(String[] args) {
		BadGrowArray b = new BadGrowArray();
    final int n = 10;
		for (int i = 0; i < n; i++)
			b.insertEnd(i); // 1 + 2 + 3 + ....  + n = O(n^2)
		for (int i = 0; i < n; i++)
			b.insertStart(i); // 1 + 2 + 3 + ....  + n = O(n^2)
		b.insert(5, 99);
		System.out.println(b);
	}

	
}

File: old/session06_GrowArray_LinkedList_Stack_Queue/GrowArray/BadGrowArray2018F.cc
class BadGrowArray {
private:
	int size;
	int* p;
public:
  BadGrowArray() : size(0), p(nullptr) {} 

	void insertEnd(int v) {  //O(n)
		int* old = p;  //O(1)
    p = new int[size + 1];  //O(1)
    for (int i = 0; i < size; i++) //O(size) = O(n)
			p[i] = old[i];
		p[size++] = v;  //O(1)
		delete [] old; //O(1)
	}
	void insertStart(int v) {
		int* old = p;
		p = new int[size+1];
		p[0] = v;
    for (int i = 1; i <= size; i++)
			p[i] = old[i-1];
		size++;
		delete [] old;		
	}
	void insert(int pos, int v) {
		int* old = p;
		p = new int[size+1];
		for (int i = 0; i < pos; i++) //O(n)
			p[i] = old[i];
		p[pos] = v;
		for (int i = pos + 1; i <= size; i++) //O(n)
			p[i] = old[i-1];
		size++;
		delete [] old;		
	}
	void removeStart() {
		for (int i = 0; i < size-1; i++)
			p[i] = p[i+1];
		size--;

	}

	void removeEnd() {
    size--;
	}
	friend ostream& operator <<(ostream& s, const BadGrowArray& b) {
		for (int i = 0; i < b.size; i++)
			s << b.p[i] << ' ';
		return s;
	}
};


int main() {
	BadGrowArray b;
	for (int i = 0; i < 10; i++)
		b.insertEnd(i); // 1 + 2 + 3 + .. + n = O(n^2)
	for (int i = 0; i < 10; i++)
		b.insertStart(i); // 1 + 2 + 3 + .. + n = O(n^2)
	for (int i = 0; i < 10; i++)
		b.insert(5, 99); // 1 + 2 + 3 + .. + n = O(n^2)
	cout << b << '\n';
	for (int i = 0; i < 20; i++)
		b.removeStart();
	cout << b << '\n';
  
}

/*
   x = [];
   for i=1:1000000
     x = [x, i];
   end
   x = zeros(1, 1000000);
   for i=1:1000000
     x(i) = i;

 */
